File: src/_cpp/TowerDefenseBase/scenes/Tutorial.cpp
//
//  Tutorial.cpp
//  JungleDefense
//
//  Created by Vladimir Tolmachev on 24.04.14.
//
//
#include "ml/SmartScene.h"
#include "ml/loadxml/xmlProperties.h"
#include "ml/ImageManager.h"
#include "ml/Language.h"
#include "Tutorial.h"
#include "resources.h"
#include "gameboard.h"
#include "GameGS.h"
#include "consts.h"
#include "UserData.h"
#include "ScoreCounter.h"
#include "configuration.h"
#include "Hero.h"
NS_CC_BEGIN;

const std::string ksAutoPosition( "autopos" );
const std::string ksParent( "parent" );
const std::string ksAsBlockingLayer( "asblockinglayer" );
const std::string ksCloseByTap( "closebytap" );
const std::string ksNextTutorial( "nexttutorial" );
const std::string ksShopGift( "shopgift" );
const std::string ksSetLevelScore( "setlevelsscore" );
const std::string ksSelectHero( "selecthero" );
const std::string ksResetHeroSkills( "resetheroskills" );
const std::string ksHerominlevel( "herominlevel" );
const int kAutoPosition( xmlLoader::kUserProperties + 1 );
const int kParent( xmlLoader::kUserProperties + 2 );
const int kAsBlockingLayer( xmlLoader::kUserProperties + 3 );
const int kCloseByTap( xmlLoader::kUserProperties + 4 );
const int kNextTutorial( xmlLoader::kUserProperties + 5 );
const int kShopGift( xmlLoader::kUserProperties + 6 );
const int kSetLevelScore( xmlLoader::kUserProperties + 7 );
const int kSelectHero( xmlLoader::kUserProperties + 8 );
const int kResetHeroSkills( xmlLoader::kUserProperties + 9 );
const int kHerominlevel( xmlLoader::kUserProperties + 10 );

/******************************************************************************/
//MARK: Tutorial
/******************************************************************************/

Tutorial::Tutorial()
{}

Tutorial::~Tutorial()
{}

bool Tutorial::init( const std::string & pathToXml )
{
	load( pathToXml );

	//assert( getParent() != nullptr );
	//{
	//	auto scene = Director::getInstance()->getRunningScene();
	//	auto smartscene = dynamic_cast<SmartScene*>(scene);
	//}

	return true;
}

bool Tutorial::setProperty( int intproperty, const std::string & value )
{
	Node * node( nullptr );
	switch( intproperty )
	{
		case xmlLoader::kPos:
			setPosition( getPosition() + strToPoint( value ) );
			break;
		case kAutoPosition:
			node = getNodeByPath( Director::getInstance()->getRunningScene(), value );
			if( node ) setPosition( getPosition() + node->getPosition() );
			break;
		case kParent:
			assert( getParent() == nullptr );
			node = getNodeByPath( Director::getInstance()->getRunningScene(), value );
			if( node ) node->addChild( this );
			break;
		case kAsBlockingLayer:
			if( strToBool( value ) )
			{
				auto scene = Director::getInstance()->getRunningScene();
				auto smartscene = dynamic_cast<SmartScene*>(scene);
				smartscene->pushLayer( this, true );
			}
			break;
		case kCloseByTap:
			if( strToBool( value ) )
			{
				auto *listener = EventListenerTouchAllAtOnce::create();
				listener->onTouchesBegan = CC_CALLBACK_2( Tutorial::touchesBegan, this );
				listener->onTouchesEnded = CC_CALLBACK_2( Tutorial::touchesEnded, this );
				Director::getInstance()->getEventDispatcher()->addEventListenerWithSceneGraphPriority( listener, this );
			}
			else
			{
				Director::getInstance()->getEventDispatcher()->removeEventListenersForTarget( this );
			}
			break;
		case kNextTutorial:
			_nextTutorial = value;
			break;
		case kShopGift:
			UserData::shared().write( k::user::ShopGift, value );
			break;
		case kSetLevelScore:
			ScoreCounter::shared().setMoney( kScoreLevel, strToInt( value ), false );
			break;
		case kSelectHero:
			UserData::shared().hero_setCurrent( strToInt( value ) );
			break;
		case kResetHeroSkills:
		{
			auto heroname = "hero" + intToStr( strToInt( value ) + 1 );
			std::vector<unsigned> skills( 5 );
			HeroExp::shared().setSkills( heroname, skills );
			break;
		}
		case kHerominlevel:
		{
			auto current = UserData::shared().hero_getCurrent();
			auto heroname = "hero" + intToStr( current + 1 );
			auto exp = HeroExp::shared().getEXP( heroname );
			auto expmin = HeroExp::shared().getHeroLevelExtValue( strToInt( value ) ) + 1;
			exp = std::max( exp, expmin );
			HeroExp::shared().setEXP( heroname, exp );
			break;
		}
		default:
			return NodeExt::setProperty( intproperty, value );
	}
	return true;
}

void Tutorial::enter()
{
	runEvent( "onenter" );
}

void Tutorial::exit()
{
	runEvent( "onexit" );
}

const std::string& Tutorial::next()const
{
	return _nextTutorial;
};

ccMenuCallback Tutorial::get_callback_by_description( const std::string & name )
{
	if( name == "confirm_tutorial_yes" )
		return std::bind( &Tutorial::cb_confirmTutorial, this, std::placeholders::_1, true );
	if( name == "confirm_tutorial_no" )
		return std::bind( &Tutorial::cb_confirmTutorial, this, std::placeholders::_1, false );
	return nullptr;
}

void Tutorial::cb_confirmTutorial( Ref*, bool use )
{
	UserData::shared().write( k::user::UseTitorial, use );
	TutorialManager::shared().close( this );
}

void Tutorial::touchesBegan( const std::vector<Touch*>& touches, Event *event )
{

}

void Tutorial::touchesEnded( const std::vector<Touch*>& touches, Event *event )
{
	TutorialManager::shared().close( this );
}


/******************************************************************************/
//MARK: TutorialManager
/******************************************************************************/

TutorialManager::TutorialManager()
{
	xmlLoader::boolProperty( ksAutoPosition, kAutoPosition );
	xmlLoader::boolProperty( ksParent, kParent );
	xmlLoader::boolProperty( ksAsBlockingLayer, kAsBlockingLayer );
	xmlLoader::boolProperty( ksCloseByTap, kCloseByTap );
	xmlLoader::boolProperty( ksNextTutorial, kNextTutorial );
	xmlLoader::boolProperty( ksShopGift, kShopGift);
	xmlLoader::boolProperty( ksSetLevelScore, kSetLevelScore );
	xmlLoader::boolProperty( ksSelectHero, kSelectHero );
	xmlLoader::boolProperty( ksResetHeroSkills, kResetHeroSkills );
	xmlLoader::boolProperty( ksHerominlevel, kHerominlevel );
}

TutorialManager::TutorialManager( TutorialManager&& )
{}

TutorialManager::TutorialManager( const TutorialManager& )
{}

TutorialManager& TutorialManager::operator=(const TutorialManager&)
{
	return *this;
}

void TutorialManager::onCreate()
{
	load();
}

bool TutorialManager::dispatch( const std::string & eventname, const ParamCollection * params )
{
	bool result( false );

	if( _current && _current->isRunning() == false )
	{
		close( _current );
		result = true;
	}

	if( !_current && _queueEvents.empty() == false )
	{
		auto pair = _queueEvents.front();
		_queueEvents.pop();
		if( dispatch(pair.first, &pair.second) )
		{
			return true;
		}
	}

	if( _current )
	{
		auto range = _eventsForClose.equal_range( eventname );
		for( auto iter = range.first; iter != range.second; ++iter )
		{
			if( iter->second == _current->getName() )
			{
				if( close( _current ) == false )
					dispatch( eventname, params );
				result = true;
				break;
			}
		}
		//if( result == false )
		//{
		//	auto range = _eventsForRun.equal_range( eventname );
		//	for( auto iter = range.first; iter != range.second; ++iter )
		//	{
		//		if( checkOpening( iter->second ) )
		//		{
		//			std::pair< std::string, ParamCollection > pair;
		//			pair.first = eventname;
		//			if( params ) pair.second = *params;
		//			_queueEvents.push( pair );
		//		}
		//	}
		//}

	}
	else
	{
		auto range = _eventsForRun.equal_range( eventname );
		for( auto iter = range.first; iter != range.second; ++iter )
		{
			std::string name = iter->second;
			if( open( name ) )
			{
				result = true;
				break;
			}
		}
	}
	return result;
}

bool TutorialManager::close( Tutorial * tutorial )
{
	if( _current && tutorial == _current )
	{
		int value = UserData::shared().get_int( "tutorial" + _current->getName() );
		UserData::shared().write( "tutorial" + _current->getName(), value + 1 );

		_current->exit();
		_current->removeFromParent();

		std::string next = _current->next();
		_current.reset( nullptr );

		if( next.empty() == false )
			open( next );

		return next.empty() == false;
	}
	else
	{
		assert( tutorial );
		tutorial->exit();
		tutorial->removeFromParent();
		return true;
	}
}

bool TutorialManager::open( const std::string & name )
{
	if( _list.find( name ) == _list.end() )
        return false;
	const TutorialInfo& info = _list[name];
	bool result( false );

	int value = UserData::shared( ).get_int( "tutorial" + name, 0 );
	if( checkOpening(name) )
	{
		UserData::shared().write( "tutorial" + name, value + 1 );

		assert( info.filename.empty() == false );
		_current = Tutorial::create( info.filename );
		_current->setName( name );
		_current->enter();
		result = true;
	}


	if( name == "lvl0_hero" )
	{
		int level = GameGS::getInstance()->getGameBoard().getCurrentLevelIndex();
		
		if( _current )
		{
			if( level == 0 && k::configuration::minLevelHero > 0 )
				result = close( _current );
			if( level == 3 && k::configuration::minLevelHero == 0 )
				result = close( _current );
		}
	}

	return result;
}

bool TutorialManager::checkOpening( const std::string tutorialname )const
{
	auto iter = _list.find( tutorialname );
	if (iter == _list.end())
		return false;
	const TutorialInfo& info = iter->second;

	bool isshow( true );
	int value = UserData::shared( ).get_int( "tutorial" + tutorialname, 0 );
	bool rejection = UserData::shared().get_bool( k::user::UseTitorial, true );
	std::string after = info.onlyaftertutorial;
	bool showprevios = after.empty() ? true : UserData::shared().get_int( "tutorial" + after ) > 0;
	isshow = isshow && showprevios;
	isshow = isshow && value < info.count;
	isshow = isshow && (rejection || info.forced);

	return isshow;
}

void TutorialManager::load()
{
	pugi::xml_document doc;
	doc.load_file( "ini/tutorial/tutorials.xml" );
	auto root = doc.root().first_child();
	auto xmllist = root.child( "list" );
	auto xmlevents = root.child( "events" );
	auto xmlrun = xmlevents.child( "run" );
	auto xmlclose = xmlevents.child( "close" );

	loadList( xmllist );
	loadEvents( xmlrun, _eventsForRun );
	loadEvents( xmlclose, _eventsForClose );
}

void TutorialManager::loadList( const pugi::xml_node & xmlnode )
{
	FOR_EACHXML( xmlnode, child )
	{
		auto dispatch = true;
		if( child.attribute( "dispatch" ) )
		{
			std::string value = child.attribute( "dispatch" ).as_string();
			dispatch = strToBool( xmlLoader::macros::parse( value ) );
		}
		if (dispatch == false)
			continue;

		TutorialInfo info;
		std::string name = child.name();
		info.filename = child.attribute( "filename" ).value();
		info.onlyaftertutorial = child.attribute( "after" ).value( );
		info.count = child.attribute( "count" ).as_int( 1 ) * 2;
		info.forced = child.attribute( "forced" ).as_bool( false );
		_list[name] = info;
	}
}

void TutorialManager::loadEvents( const pugi::xml_node & xmlnode, std::multimap<std::string, std::string> & events )
{
	FOR_EACHXML( xmlnode, child )
	{
		std::string eventname = child.name();
		std::string tutorial = child.attribute( "value" ).value();
		events.insert( std::pair<std::string, std::string>( eventname, tutorial ) );
	}
}


NS_CC_END;

File: src/_eu/scenes/Tutorial.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

/**TESTED**/
//Define namespace
var EU = EU || {};

EU.ksAutoPosition = "autopos";
EU.ksParent = "parent";
EU.ksAsBlockingLayer = "asblockinglayer";
EU.ksCloseByTap = "closebytap";
EU.ksNextTutorial = "nexttutorial";
EU.ksShopGift = "shopgift";
EU.ksSetLevelScore = "setlevelsscore";
EU.ksSelectHero = "selecthero";
EU.ksResetHeroSkills = "resetheroskills";
EU.ksHerominlevel = "herominlevel";
EU.kAutoPosition = EU.xmlKey.UserProperties.int + 1;
EU.kParent = EU.xmlKey.UserProperties.int + 2;
EU.kAsBlockingLayer = EU.xmlKey.UserProperties.int + 3;
EU.kCloseByTap = EU.xmlKey.UserProperties.int + 4;
EU.kNextTutorial = EU.xmlKey.UserProperties.int + 5;
EU.kShopGift = EU.xmlKey.UserProperties.int + 6;
EU.kSetLevelScore = EU.xmlKey.UserProperties.int + 7;
EU.kSelectHero = EU.xmlKey.UserProperties.int + 8;
EU.kResetHeroSkills = EU.xmlKey.UserProperties.int + 9;
EU.kHerominlevel = EU.xmlKey.UserProperties.int + 10;

/******************************************************************************/
//MARK: Tutorial
/******************************************************************************/
EU.Tutorial = cc.Layer.extend(
{
    /** @type {String} */ _nextTutorial : null,

    ctor: function( /**@type {String} */ pathToXml )
    {
        this._super();
        this._nextTutorial = "";
        this.initExt();
        this.load_str( pathToXml );

        //EU.assert( this.getParent() != null );
        //{
        //	var scene = cc.director.getRunningScene();
        //	var smartscene = dynamic_cast<SmartScene*>(scene);
        //}

        return true;
    },


    enter: function()
    {
        this.runEvent( "onenter" );
    },

    exit: function()
    {
        this.runEvent( "onexit" );
    },

    next: function()
    {
        return this._nextTutorial;
    },

    cb_confirmTutorial: function( use )
    {
        EU.UserData.write( EU.k.UseTitorial, use );
        EU.TutorialManager.close( this );
    },

    touchesBegan: function( touches, event )
    {

    },

    touchesEnded: function(  touches, event )
    {
        EU.TutorialManager.close( this );
    }

});

EU.NodeExt.call(EU.Tutorial.prototype);
EU.Tutorial.prototype.setProperty_int = function( /**@type {Integer} */ intproperty, /**@type {String} */ value ) {
    var node = null;
    switch( intproperty )
    {
        case EU.xmlKey.Pos.int:
            this.setPosition( cc.pAdd(this.getPosition(), EU.Common.strToPoint( value ) ) );
            break;
        case EU.kAutoPosition:
            node = EU.Common.getNodeByPath( cc.director.getRunningScene(), value );
            if( node ) this.setPosition( cc.pAdd(this.getPosition() ,node.getPosition()) );
            break;
        case EU.kParent:
            EU.assert( this.getParent() == null );
            node = EU.Common.getNodeByPath( cc.director.getRunningScene(), value );
            if( node ) node.addChild( this );
            break;
        case EU.kAsBlockingLayer:
            if( EU.Common.strToBool( value ) )
            {
                var scene = cc.director.getRunningScene();
                var smartscene = scene;
                if (!smartscene.__SmartScene) smartscene = null;
                smartscene.pushLayer( this, true );
            }
            break;
        case EU.kCloseByTap:
            if( EU.Common.strToBool( value ) )
            {
                var listener = cc.EventListener.create({
                    event: cc.EventListener.TOUCH_ALL_AT_ONCE,
                    onTouchBegan: this.touchesBegan.bind( this ),
                    onTouchesEnded: this.touchesEnded.bind( this )
                });
                cc.eventManager.addListener(listener, this);
            }
            else
            {
                cc.eventManager.removeListeners(this);
            }
            break;
        case EU.kNextTutorial:
            this._nextTutorial = value;
            break;
        case EU.kShopGift:
            EU.UserData.write( EU.k.user.ShopGift, value );
            break;
        case EU.kSetLevelScore:
            EU.ScoreCounter.setMoney( EU.kScoreLevel, EU.Common.strToInt( value ), false );
            break;
        case EU.kSelectHero:
            EU.UserData.hero_setCurrent( EU.Common.strToInt( value ) );
            break;
        case EU.kResetHeroSkills:
        {
            var heroname = "hero" + ( EU.Common.strToInt( value ) + 1 );
            var skills = [];
            //TODO: EU.HeroExp.setSkills( heroname, skills );
            break;
        }
        case EU.kHerominlevel:
        {
            //TODO:
            //var current = EU.UserData.hero_getCurrent();
            //var heroname = "hero" + ( current + 1 );
            //var exp = EU.HeroExp.getEXP( heroname );
            //var expmin = EU.HeroExp.getHeroLevelExtValue( EU.Common.strToInt( value ) ) + 1;
            //exp = Math.max( exp, expmin );
            //EU.HeroExp.setEXP( heroname, exp );
            break;
        }
        default:
            return false;
    }
    return true;
};
EU.Tutorial.prototype.get_callback_by_description = function( /**@type {String} */ name ) {
    var self = this;
    if( name == "confirm_tutorial_yes" )
        return this.cb_confirmTutorial.bind( this, true );
    if( name == "confirm_tutorial_no" )
        return this.cb_confirmTutorial.bind( this, false );
    return null;
};



/******************************************************************************/
//MARK: TutorialManager
/******************************************************************************/
EU.TutorialManager = {


    //var dispatch(  var  eventname,  ParamCollection * params = null );
    /** @type {this} */ _current : null,
    TutorialInfo: function(filename, onlyaftertutorial, count, forced)
    {
        this.filename = filename || "";
        this.onlyaftertutorial = onlyaftertutorial || "";
        this.count = count || 1;
        this.forced = forced || false;
    },
    /** @type {map<string, EU.TutorialManager.TutorialInfo>} */ _list : null,
    /*
     Events for run tutorial.
     * first - eventname
     * second - tutorial name
     */
    /** @type {map<string, Array<string>>} */ _eventsForRun : null,
    /*
     Events for close tutorial.
     * first - eventname
     * second - tutorial name
     */
    /** @type {map<string, Array<string>>} */ _eventsForClose : null,

    /** @type {Array<[ string, ParamCollection ] >} */ _queueEvents : null,


    onCreate: function()
    {

        this._list = {};
        this._eventsForRun = {};
        this._eventsForClose = {};
        this._queueEvents = [];
        EU.xmlLoader.bookProperty( EU.ksAutoPosition, EU.kAutoPosition );
        EU.xmlLoader.bookProperty( EU.ksParent, EU.kParent );
        EU.xmlLoader.bookProperty( EU.ksAsBlockingLayer, EU.kAsBlockingLayer );
        EU.xmlLoader.bookProperty( EU.ksCloseByTap, EU.kCloseByTap );
        EU.xmlLoader.bookProperty( EU.ksNextTutorial, EU.kNextTutorial );
        EU.xmlLoader.bookProperty( EU.ksShopGift, EU.kShopGift);
        EU.xmlLoader.bookProperty( EU.ksSetLevelScore, EU.kSetLevelScore );
        EU.xmlLoader.bookProperty( EU.ksSelectHero, EU.kSelectHero );
        EU.xmlLoader.bookProperty( EU.ksResetHeroSkills, EU.kResetHeroSkills );
        EU.xmlLoader.bookProperty( EU.ksHerominlevel, EU.kHerominlevel );

        this.load();
    },

    dispatch: function( eventname,  params )
    {
        var result = false;

        if( this._current && this._current.isRunning() == false )
        {
            this.close( this._current );
            result = true;
        }

        if( !this._current && this._queueEvents.length > 0 )
        {
            var pair = this._queueEvents.shift();
            if( this.dispatch(pair[0], pair[1]) )
            {
                return true;
            }
        }

        if( this._current )
        {
            var range = this._eventsForClose[eventname];
            for (var i = 0; range != null && i < range.length; i++) {
                var iter = range[i];
                if( iter == this._current.getName() )
                {
                    if( this.close( this._current ) == false )
                        this.dispatch( eventname, params );
                    result = true;
                    break;
                }
            }
            //if( result == false )
            //{
            //	var range = this._eventsForRun.equal_range( eventname );
            //	for( var iter = range.first; iter != range.second; ++iter )
            //	{
            //		if( this.checkOpening( iter.second ) )
            //		{
            //			pair< string, ParamCollection > pair;
            //			pair.first = eventname;
            //			if( params ) pair.second = *params;
            //			this._queueEvents.push( pair );
            //		}
            //	}
            //}

        }
        else
        {
            var range = this._eventsForRun[ eventname ];
            for (var i = 0; range != null && i < range.length; i++) {
                var name = range[i];
                if( this.open( name ) )
                {
                    result = true;
                    break;
                }
            }
        }
        return result;
    },

    close: function( tutorial )
    {
        if( this._current && tutorial == this._current )
        {
            var value = EU.UserData.get_int( "tutorial" + this._current.getName() );
            EU.UserData.write( "tutorial" + this._current.getName(), value + 1 );

            this._current.exit();
            this._current.removeFromParent(true);

            var next = this._current.next();
            this._current = null;
            var result = cc.isString(next) && next.length > 0;
            if( result )
                this.open( next );

            return result;
        }
        else
        {
            EU.assert( tutorial );
            tutorial.exit();
            tutorial.removeFromParent(true);
            return true;
        }
    },

    open: function( /**@type {String} */ name )
    {
        if( this._list[name] == null )
            return false;
        var info = this._list[name];
        var result = false;

        var value = EU.UserData.get_int( "tutorial" + name, 0 );
        if( this.checkOpening(name) )
        {
            EU.UserData.write( "tutorial" + name, value + 1 );

            EU.assert( info.filename.length > 0 );
            this._current = new EU.Tutorial( info.filename );
            this._current.setName( name );
            this._current.enter();
            result = true;
        }

        if( name == "lvl0_hero" )
        {
           var level = EU.GameGSInstance.getGameBoard().getCurrentLevelIndex();

            if( this._current )
            {
                if( level == 0 && EU.k.minLevelHero > 0 )
                    result = this.close( this._current );
                if( level == 3 && EU.k.minLevelHero == 0 )
                    result = this.close( this._current );
            }
        }

        return result;
    },

    checkOpening: function( /**@type {String} */ tutorialname )
    {
        var iter = this._list[tutorialname];
        if (iter == null)
            return false;
        var info = iter;

        var isshow = true ;
        var value = EU.UserData.get_int( "tutorial" + tutorialname, 0 );
        var rejection = EU.UserData.get_bool( EU.k.UseTitorial, true );
        var after = info.onlyaftertutorial;
        var showprevious = after.length == 0 ? true : EU.UserData.get_int( "tutorial" + after ) > 0;
        isshow = isshow && showprevious;
        isshow = isshow && value < info.count;
        isshow = isshow && (rejection || info.forced);

        return isshow;
    },

    load: function()
    {
        /**@type {Element} */
        var doc = null;
        EU.pugixml.readXml("ini/tutorial/tutorials.xml", function(error, data) {
            doc = data;
        }, this);
        var root = doc.firstElementChild;
        var xmllist = root.getElementsByTagName( "list" )[0];
        var xmlevents = root.getElementsByTagName( "events" )[0];
        var xmlrun = xmlevents.getElementsByTagName( "run" )[0];
        var xmlclose = xmlevents.getElementsByTagName( "close" )[0];

        this.loadList( xmllist );
        this.loadEvents( xmlrun, this._eventsForRun );
        this.loadEvents( xmlclose, this._eventsForClose );
    },

    loadList: function( /**@type {Element} */ xmlnode )
    {
        for (var i = 0; i < xmlnode.children.length; i++) {
            var child = xmlnode.children[i];
            var dispatch = true;
            if( child.getAttribute( "dispatch" ) )
            {
                var value = child.getAttribute( "dispatch" );
                dispatch = EU.Common.strToBool( EU.xmlLoader.macros.parse( value ) );
            }
            if (dispatch == false)
                continue;

            var info = new EU.TutorialManager.TutorialInfo();
            var name = child.tagName;
            info.filename = EU.asObject(child.getAttribute( "filename" ), "");
            info.onlyaftertutorial = EU.asObject(child.getAttribute( "after" ), "");
            info.count = parseInt(EU.asObject(child.getAttribute( "count" ), "1" )) * 2;
            info.forced = EU.Common.strToBool(EU.asObject(child.getAttribute( "forced" ), "false" ));
            this._list[name] = info;
        }
    },

    loadEvents: function( xmlnode, /**map<string, Array<string>>*/ events )
    {
        for (var i = 0; i < xmlnode.children.length; i++) {
            var child = xmlnode.children[i];
            var eventname = child.tagName;
            var tutorial = EU.asObject(child.getAttribute( "value" ), "");
            if (events[eventname] == null) events[eventname] = [];
            events[eventname].push( tutorial );
        }
    }

};

EU.TutorialManager.onCreate();